### PR TITLE
allow both aip and parallel to be enabled together

### PR DESCRIPTION
**Pull request by simonafk**

### CODE REVIEW

File: tests/test_flow/test_persistence_aip.py [hunk 1]
```diff
@@ -2,30 +2,50 @@
...
 import bionic as bn
 
-
 # This is detected by pytest and applied to all the tests in this module.
-pytestmark = [pytest.mark.no_parallel]
+pytestmark = pytest.mark.needs_aip
 
 
-@pytest.mark.needs_aip
 def test_aip_jobs(aip_builder):
```
Comment (jqmp) [inline tests/test_flow/test_persistence_aip.py:9]: Do we have any tests that check that the `aip_task_config` tasks were actually sent to AIP (or to a fake AIP)? If not, could we add a TODO for that?

On a similar note: are two test cases (this one and `test_logging`) enough? Are there any other tests we should come back and add? For example, can/should we have a case where an AIP task and a subprocess task are running in parallel?

File: tests/test_flow/fakes.py [hunk 2]
```diff
@@ -171,28 +117,65 @@ def pipe(self, url, content_bytes):
...
+    filter_all_logs = FilterAllLogs()
+    try:
+        logger.addFilter(filter_all_logs)
+        yield
+    finally:
+        logger.removeFilter(filter_all_logs)
+
+
+@contextmanager
+def temporarily_clean_cache_dir(tmp_path):
```
Comment (jqmp) [inline tests/test_flow/fakes.py:136]: Is it possible for another process to be running that's accessing the same directory? For example, if we're using parallel execution?

File: bionic/core/flow_execution.py [hunk 3]
```diff
@@ -398,28 +404,14 @@ def _mark_entry_active(self, entry):
...
-            # and newer tasks can potentially conflict writing to the
-            # same cache file.
-            task_keys = finished_future.result()
-            for task_key in task_keys:
+        # With AIP execution, we can have one task fail while others are
+        # running. If we don't wait till those tasks are done, the user might
+        # perform another `get` operation that can spawn the same Task again.
+        # Apart from wasted resource, the other problem is that the older and
+        # newer tasks can potentially conflict writing to the same cache file.
+        future_results = [future.result() for future in futures]
```
Comment (jqmp) [inline bionic/core/flow_execution.py:412]: Wait, does this mean we now always block until every in-progress job is finished? That sounds like it will make parallel processing worse. What if one job takes much longer than before?

Is there another way we can resolve the duplicate-AIP-job problem that you mention here?

File: tests/test_flow/fakes.py [hunk 4]
```diff
@@ -171,28 +117,65 @@ def pipe(self, url, content_bytes):
...
+            "bionic.aip.main",
+        ]
+
+        input_uri = body["trainingInput"]["args"][3]
+        assert self.gcs_fs.exists(input_uri)
+
+        # AIP execution do not have access to local disk cache and do not send
+        # back logs to local bionic instance. Emulate this behavior while
+        # running the AIP job.
+        with temporarily_filter_all_logs(), temporarily_clean_cache_dir(self.tmp_path):
```
Comment (jqmp) [inline tests/test_flow/fakes.py:169]: It looks like we just run this job synchronously. Does this mean it's impossible for us to have a test case where multiple fake AIP jobs are in progress at once?

**Review by jqmp — COMMENTED** [review]
Overall looks reasonable, but there are some edge cases around concurrency that I'm not sure about. In particular, removing the use of `concurrent.futures.wait` seems like it might be 
a problem.

I also asked some more questions about test cases, which currently look pretty minimal; I know that's not the point of this PR, but maybe we could add some TODOs?

File: bionic/core/flow_execution.py (hunk 3 above)
Comment (simonafk) [reply bionic/core/flow_execution.py:412]: Hmm, I used the `.result()` here because I was looking at the TODO. After reading the comments in the PR and thinking about concurrency more, I will need to rethink about the implementation.

File: tests/test_flow/fakes.py (hunk 4 above)
Comment (simonafk) [reply tests/test_flow/fakes.py:169]: We are calling `executor.submit` one at a time. So I think if we are running the job synchronously during creation, we should not have multiple fake AIP jobs running concurrently. But as you mentioned, it may be possible for a parallel executor to be running at the same time.

File: tests/test_flow/fakes.py (hunk 2 above)
Comment (simonafk) [reply tests/test_flow/fakes.py:136]: Good catch.

One idea I have is to make the parallel executor synchronous for tests.

File: bionic/aip/task.py [hunk 5]
```diff
@@ -49,11 +48,12 @@ class Config:
 
 @attr.s(auto_attribs=True, frozen=True)
 class Task:
+    # This task object will be serialized and sent to AIP. Hence, all entities
+    # here must be serializable as well.
     name: str
     function: Callable
     config: Config
     task_config: TaskConfig
-    gcs_fs: Any  # AbstractFileSystem-like object, not necessarily a subclass.
```
[the remark this answers is not in the document]
Comment (simonafk) [reply bionic/aip/task.py:56]: Removing it because it does not seem to be necessary. We are already storing and using the `gcs_fs` in `TaskCompletionRunner._core`.

File: tests/test_flow/test_persistence_aip.py (hunk 1 above)
Comment (simonafk) [reply tests/test_flow/test_persistence_aip.py:12]: Expanded test to have functions that can run in parallel as well.

File: tests/test_flow/fakes.py (hunk 2 above)
Comment (simonafk) [reply tests/test_flow/fakes.py]: Switched to a different approach for file isolation for fake AIP. The new approach will extract the `TaskCompletionRunner` object from fake GCS, modify the cache directory path, and save it back into fake GCS.

File: bionic/core/flow_execution.py (hunk 3 above)
Comment (simonafk) [reply bionic/core/flow_execution.py]: The AIP executor now spawns a separate thread for querying the job status, returning a concurrent Future object. Hence, we can use the `wait()` function from Python's concurrent future library.

Also, if an exception is thrown, then bionic will wait for all the futures to complete and be saved to disk if possible.

File: bionic/core/flow_execution.py [hunk 6]
```diff
@@ -398,28 +404,14 @@ def _mark_entry_active(self, entry):
...
-            # from wasted resource, the other problem is that the older
-            # and newer tasks can potentially conflict writing to the
-            # same cache file.
-            task_keys = finished_future.result()
-            for task_key in task_keys:
+        # With AIP execution, we can have one task fail while others are
+        # running. If we don't wait till those tasks are done, the user might
+        # perform another `get` operation that can spawn the same Task again.
+        # Apart from wasted resource, the other problem is that the older and
+        # newer tasks can potentially conflict writing to the same cache file.
```
[the remark this answers is not in the document]
Comment (simonafk) [reply bionic/core/flow_execution.py]: Hmm, what should this be changed into?

File: tests/test_flow/conftest.py [hunk 7]
```diff
@@ -53,10 +53,9 @@ def builder(parallel_execution_enabled, tmp_path):
 
 # This is a different multiprocessing manager than the one we use in
 # ExternalProcessLoggingManager. This one is responsible for sharing test
-# objects between processes. It's only used in make_counter, make_list and
-# make_dict fixtures. The purpose here is to not pollute the manager that bionic
-# uses. I also don't want to replace the manager that bionic creates with a test
-# one.
+# objects between processes. It's only used in make_counter, make_list fixtures.
```
Comment (jqmp) [inline tests/test_flow/conftest.py:56]: nit: s/used in make_counter, make_list/used in the make_counter and make_list/

File: bionic/core/flow_execution.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jqmp) [reply bionic/core/flow_execution.py]: I think we can remove the last sentence, since AFAIK this problem doesn't actually exist.

File: tests/test_flow/test_persistence_aip.py [hunk 8]
```diff
@@ -1,31 +1,142 @@
...
+
+    y1_counter = make_counter()
+    y2_counter = make_counter()
+    y4_counter = make_counter()
+    y5_counter = make_counter()
+
+    @builder
+    @y1_counter
+    def y1(x):
+        time.sleep(20)
```
[the remark this answers is not in the document]
Comment (simonafk) [reply tests/test_flow/test_persistence_aip.py:86]: I have been hesitant about making tests that run only on fake AIP. But now I think it can be worthwhile for testing out the exception catching logic.

File: bionic/core/flow_execution.py [hunk 9]
```diff
@@ -401,31 +408,44 @@ def _mark_entry_active(self, entry):
...
             # same cache file.
-            task_keys = finished_future.result()
-            for task_key in task_keys:
-                entry = self._entries_by_task_key[task_key]
-                entry.state.sync_after_remote_computation()
-                assert self._mark_entry_completed_if_possible(entry)
+            logger.error(
+                oneline(
+                    f"""
+                    While computing remotely, an exception was thrown:
```
[the remark this answers is not in the document]
Comment (simonafk) [reply bionic/core/flow_execution.py:430]: The information about the task being computed is in the exception itself.

File: bionic/aip/task.py [hunk 10]
```diff
@@ -124,8 +124,28 @@ def submit(self) -> Future:
...
+            return pickle.load(f)
+
+    def _get_state_and_error(self, aip_client):
+        request = (
+            aip_client.projects()
+            .jobs()
+            .get(name=f"projects/{self.config.project_name}/jobs/{self.job_id()}")
         )
+        response = request.execute()
+        state, error = State[response["state"]], response.get("errorMessage", "")
```
[the remark this answers is not in the document]
Comment (simonafk) [reply bionic/aip/task.py:150]: `""` is convenient for now since we are just printing it out if there is an error.

File: bionic/core/flow_execution.py [hunk 11]
```diff
@@ -401,31 +408,44 @@ def _mark_entry_active(self, entry):
...
+            for future in completed_futures:
+                try:
+                    task_keys = future.result()
+                    for task_key in task_keys:
+                        entry = self._entries_by_task_key[task_key]
+                        entry.state.sync_after_remote_computation()
+                        assert self._mark_entry_completed_if_possible(entry)
+                except Exception as e:
+                    if e != exception:
+                        logger.error(f"Another exception was thrown: {e}")
```
[the remark this answers is not in the document]
Comment (simonafk) [reply bionic/core/flow_execution.py:446]: `wait(futures, return_when=ALL_COMPLETED)` should ensure that all the futures are completed.

[the remark this answers is not in the document]
Comment (jqmp) [reply bionic/core/flow_execution.py]: Oh, you're right! But in that case, wouldn't we want to print all the extra errors, not just the first one?

File: tests/test_flow/test_persistence_aip.py [hunk 12]
```diff
@@ -1,31 +1,142 @@
...
 
     @builder
-    def c(a, b):
-        return a - b  # 6 (2 * 3) - 5 (2 + 3)
+    def total(y1, y2, y3, y4, y5):
+        return y1 + y2 + y3 + y4 + y5
+
+    if fake_aip is not None:
+        fake_aip.countdown = 0
+        fake_aip.succeed = False
```
[the remark this answers is not in the document]
Comment (simonafk) [reply tests/test_flow/test_persistence_aip.py]: Modified the fake AIP so that the final job state depends on whether an exception is thrown.

File: tests/test_flow/test_persistence_aip.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (simonafk) [reply tests/test_flow/test_persistence_aip.py]: I like the `wait_for_step` idea. Added a multiprocess barrier for blocking the entity functions.

File: tests/test_flow/fakes.py [hunk 13]
```diff
@@ -171,28 +119,80 @@ def pipe(self, url, content_bytes):
...
+        yield
+    finally:
+        logger.removeFilter(filter_all_logs)
+
+
+class FakeAipClient:
+    def __init__(self, gcs_fs, tmp_path):
+        self._gcs_fs = gcs_fs
+        self._tmp_path = tmp_path
+        self.countdown: int = 1
```
[the remark this answers is not in the document]
Comment (simonafk) [reply tests/test_flow/fakes.py]: The `countdown` was used to simulate AIP taking some time to complete the job. I have updated the tests to no longer require the countdown.

File: bionic/core/flow_execution.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (simonafk) [reply bionic/core/flow_execution.py]: Yeah, that is the purpose of the loop inside the `except`. It checks for any exceptions all of the completed futures and logs them.

[the remark this answers is not in the document]
Comment (jqmp) [reply bionic/core/flow_execution.py]: Sorry, you're totally right. Looks like I had a persistent misconception about this piece of code.

BTW, I like the streamlined version of this you have now.

[the remark this answers is not in the document]
Comment (simonafk) [reply bionic/core/flow_execution.py]: Also, added a check in the test to verify that bionic is capable of logging multiple exceptions.

File: bionic/core/flow_execution.py [hunk 14]
```diff
@@ -401,31 +408,44 @@ def _mark_entry_active(self, entry):
...
             # `get` operation that can spawn the same Task again. Apart
             # from wasted resource, the other problem is that the older
             # and newer tasks can potentially conflict writing to the
             # same cache file.
-            task_keys = finished_future.result()
-            for task_key in task_keys:
-                entry = self._entries_by_task_key[task_key]
-                entry.state.sync_after_remote_computation()
-                assert self._mark_entry_completed_if_possible(entry)
+            logger.error(
```
[the remark this answers is not in the document]
Comment (simonafk) [reply bionic/core/flow_execution.py:427]: Good point about the `logger.exception`. However, I was not able to retrieve the exception object from pytest's `caplog`. Using `logger.error` for now and printing out the exception in the message.

File: bionic/core/flow_execution.py [hunk 15]
```diff
@@ -401,31 +408,44 @@ def _mark_entry_active(self, entry):
...
+                oneline(
+                    f"""
+                    While computing remotely, an exception was thrown:
+                    {exception}. Waiting to for all other remote computation
+                    to complete.
+                    """
+                )
+            )
+            completed_futures, _ = wait(futures, return_when=ALL_COMPLETED)
```
[the remark this answers is not in the document]
Comment (simonafk) [reply bionic/core/flow_execution.py]: Hmm, reusing `finished_futures` here to reduce confusion.

File: bionic/core/flow_execution.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (simonafk) [reply bionic/core/flow_execution.py]: Yes. Added a test to verify that the entity is in the log messages.

File: bionic/core/flow_execution.py [hunk 16]
```diff
@@ -401,31 +408,44 @@ def _mark_entry_active(self, entry):
...
-            # those tasks are done, the user might perform another
+        finished_futures, _ = wait(futures, return_when=FIRST_COMPLETED)
+        try:
+            for finished_future in finished_futures:
+                task_keys = finished_future.result()
+                for task_key in task_keys:
+                    entry = self._entries_by_task_key[task_key]
+                    entry.state.sync_after_remote_computation()
+                    assert self._mark_entry_completed_if_possible(entry)
+        except Exception as exception:
```
[the remark this answers is not in the document]
Comment (simonafk) [reply bionic/core/flow_execution.py:419]: I was wondering if we need `sync_after_remote_computation` here. A successfully completed parallel execution would save its results to disk, and a successfully completed AIP execution would save its result to GCS. The next flow execution should read the data from disk or GCS.

`sync_after_remote_computation` would also ensure that the results of AIP execution would be saved to local disk.

File: bionic/core/flow_execution.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (jqmp) [reply bionic/core/flow_execution.py:427]: I believe you can get the exception object from `caplog` with `record.exc_info[1]`.

My thinking here is, it's useful to print the traceback here because it allows the user to get started on debugging while they wait for the remaining tasks to complete. (I'm less sure about the loop below, though -- it might not be necessary to print a traceback for all the other exceptions too.)

File: bionic/core/flow_execution.py [hunk 17]
```diff
@@ -401,31 +408,48 @@ def _mark_entry_active(self, entry):
...
+            logger.error(
+                oneline(
+                    f"""
+                    Encountered an error while doing remote computation:
+                    {exception!r}. Waiting for all other remote computation to
+                    complete...
+                    """
+                )
+            )
+            finished_futures, _ = wait(futures, return_when=ALL_COMPLETED)
```
Comment (jqmp) [inline bionic/core/flow_execution.py:436]: There could be futures that failed in the first batch which we haven't had a chance to report, right?

File: bionic/core/flow_execution.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jqmp) [reply bionic/core/flow_execution.py:419]: Bumping this: what are your thoughts here?

**Review by jqmp — APPROVED** [review]
I have a couple more nits about error reporting, but otherwise this looks good!

File: bionic/core/flow_execution.py (hunk 17 above)
Comment (simonafk) [reply bionic/core/flow_execution.py:436]: Yes. The `wait(...)` here should return the results of all futures, including the one that caused the exception. Hence, the next loop should be able to capture all the exceptions.

File: bionic/aip/task.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (simonafk) [reply bionic/aip/task.py]: From the documentation, it seems that `errorMessage` is only useful in the case of failure or cancellation. Anyways, switched to return `None` by default and used `str(...)` when printing out the error.

File: tests/test_flow/test_persistence_aip.py [hunk 18]
```diff
@@ -1,31 +1,142 @@
...
+        return x1 + x2 + x3 + x4 + x5 + 5
+
+    @builder
+    def total(y1, y2, y3, y4, y5):
+        return y1 + y2 + y3 + y4 + y5
+
+    assert builder.build().get("total") == 90
+
+
+# Test the scenario where an AIP job fail. Bionic should wait for other parallel
```
[the remark this answers is not in the document]
Comment (simonafk) [reply tests/test_flow/test_persistence_aip.py]: Test moved to `test_executor.py` and the comments have been rewritten.

File: bionic/core/flow_execution.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (simonafk) [reply bionic/core/flow_execution.py:425]: Good point. Also, from my testing, it looks like `exc_text` shows the stack trace from the parallel executor. Will fix this and the test in a separate PR.

File: bionic/core/flow_execution.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (simonafk) [reply bionic/core/flow_execution.py:419]: If `sync_after_remote_computation` fails, I do not think it makes sense to progress further on the computation? I think the best course of action is the same, which is to wait for all the other entities to finish computation.

File: bionic/core/flow_execution.py (hunk 17 above)
Comment (jqmp) [reply bionic/core/flow_execution.py:434]: Ah, right -- got it!

File: bionic/core/flow_execution.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jqmp) [reply bionic/core/flow_execution.py:419]: Sounds reasonable to me.
